### PR TITLE
RSDK-1130 unwrap camera for gostream ingestion

### PR DESCRIPTION
**Pull request by SeanAvery**
### Description
This PR fixes how we pass the `mediaSource` into gostream `streamMediaSource`. Instead of handing off a wrapped camera, we now extract the videoSource directly.

This is accomplished. by extending `SourceFromCamera` to handle all camera types (`reconfigurableCamera`, `CameraWaitGroup`)

### Testing

Run Server in debug mode:
```
./web/cmd/server/main.go -debug --config /etc/viam.json
```

In app.viam.com -> control -> start camera stream

Notice how the log `no properties found for media; will assume empty` no longer appears.

___

_Warning:_ CameraWaitGroup has been moved from `videosource` to `camera` package

### CODE REVIEW

**Review by bhaney — APPROVED** [review]
This all makes sense to me- thanks for adding the test!

Just make sure to run `make build lint` before committing

File: components/camera/camera.go
Comment (bazile-clyde) [inline components/camera/camera.go:234]: You can do `SourceFromCamera(asWaitGroup.Cam)` here per my comment above.

File: components/camera/camera.go
Comment (bazile-clyde) [inline components/camera/camera.go:233]: We don't know if the underlying camera (`asWaitGroup.Cam` here) will always be a `reconfigurableCamera` so we don't need to call `utils.UnwrapProxy` here. If it _is_, then a recursive call to `SourceFromCamera(asWaitGroup.Cam)` will unwrap it making this redundant. If it is _not_ then this call will do nothing and is essentially a no-op.

File: components/camera/camera_waitgroup.go
Comment (bazile-clyde) [inline components/camera/camera_waitgroup.go:16]: Why is `ActiveBackgroundWorkers` exported now?

File: robot/web/web.go
Comment (bazile-clyde) [inline robot/web/web.go:632]: `if err != nil` that means for some reason we can't get a source from our camera. I'm not sure if that should be an error but it should at least be a warning so we aren't ignoring potential errors. That is
```
if cam, ok := source.(camera.Camera); ok {
    src, err := camera.SourceFromCamera(cam)
    if err != nil {
        warn(err, ...)	// or return error
        break // do not accidentally assign source = nil 	
    }
    source = src
}

```

**Review by bazile-clyde — CHANGES_REQUESTED** [review]

File: components/camera/camera_waitgroup.go
Comment (SeanAvery) [reply components/camera/camera_waitgroup.go:16]: I moved CameraWaitGroup to camera package, so ActiveBacgroundWorers needs to be public for access in videosource.

See webcam.go diff.

Comment (edaniels) [reply components/camera/camera_waitgroup.go:16]: can we find a way to not have to do that? This feels like a bit of a footgun now based on how someone else may incorrectly use this encapsulated concept

File: components/camera/camera_test.go
Comment (bazile-clyde) [inline components/camera/camera_test.go:242]: You shouldn't need to pull out the `Cam` here. If you do you aren't testing that this function actually works with a `CameraWaitGroup` but are retesting that it works with a reconfigurable camera since `Cam = reconfCam`

File: components/camera/camera_waitgroup.go
Comment (SeanAvery) [inline components/camera/camera_waitgroup.go:14]: Renamed to avoid stutter error from linter.

```
exported: type name will be used as camera.CameraWaitGroup by other packages, and that stutters; consider calling this WaitGroup (revive)
```

File: components/camera/camera_test.go
Comment (bazile-clyde) [inline components/camera/camera_test.go:234]: You should probably test that `SourceFromCamera` returns `actualCam` here and below.

File: components/camera/camera_waitgroup.go
Comment (SeanAvery) [reply components/camera/camera_waitgroup.go:20]: Just circling back on this PR. 

Semantically, I think it makes sense for `WaitGroupCamera` to be in the `camera` package instead of `videosource`.

It also could be useful for future use cases - if more child routines are needed off the camera component.

@bazile-clyde What do you think? Any ideas on how to keep WaitGroup private?

I added a warning comment above the definition.

Comment (bazile-clyde) [reply components/camera/camera_waitgroup.go:20]: You could move the retry logic into the `CamerWaitGroup` struct since it handles the `sync.WaitGroup` stuff. The `DoCommand` handles arbitrary data. So maybe the objects the retry logic needs could go in there. Something like

```
func (c *CameraWaitGroup) DoCommand(ctx context.Context, cmd map[string]interface{}) (map[string]interface{}, error) {
	if _, ok := cmd["reconnection"]; ok {
		return c.doReconnection(ctx, cmd)	
	}
	return c.cam.DoCommand(ctx, cmd)
}

```
could work. You'd then create a map of the args you need in `NewWebcamSource`, create a `WaitGroup` then call the `DoCommand` on it before returning. The reconnection logic would then be completely encapsulated in this class. 

Furthermore, you could make `WaitGroup` fulfill the `ProxyType` interface to make this struct easier to unwrap with `UnwrapProxy`.

@edaniels Let me know if this solution abuses what the `DoCommand` was created for. Seems pretty generic.

Comment (edaniels) [reply components/camera/camera_waitgroup.go:20]: Just saw this. I think I ended up doing something similar but also removed some reconfigurable logic that was causing issues in https://github.com/viamrobotics/rdk/pull/1883

Comment (bazile-clyde) [reply components/camera/camera_waitgroup.go:20]: Cool. @SeanAvery you could wait until that PR is merged, rebase, and then fix this problem. It should be a lot simpler.

**Pull request closed without merging**
